### PR TITLE
Add support for ActionCoroutine like Unity

### LINKED ISSUE #1471: Add support for ActionCoroutine like Unity
I referred to the source code [here](https://github.com/whitglint/cccoroutine)

C++ compiler with Coroutines TS support
- VC++ : /await (Configuration Properties - C/C++ - Command Lines)
- Android : -fcoroutines-ts -stdlib=libc++ (add cppFlags in build.gradle)
- other platforms : I couldn't test it

Is it possible to test on another platform?
<br>

<details>
    <summary>ActionCoroutine.h</summary>

```c++
#ifndef __ACTION_CCCOROUTINE_ACTION_H__
#define __ACTION_CCCOROUTINE_ACTION_H__

#include "2d/Action.h"
#include <experimental/coroutine>
#include <memory>
#include <utility>

NS_AX_BEGIN

class Coroutine
{
public:
    class promise_type;
    using handle = std::experimental::coroutine_handle<promise_type>;
    class promise_type
    {
    public:
        virtual ~promise_type()
        {
            AX_SAFE_RELEASE(_currentAction);
        }

        Action* currentAction() const noexcept
        {
            return _currentAction;
        }

        auto final_suspend() const noexcept
        {
            return std::experimental::suspend_always{};
        }

        auto get_return_object() noexcept
        {
            return Coroutine(handle::from_promise(*this));
        }

        auto initial_suspend() const noexcept
        {
            return std::experimental::suspend_always{};
        }

        auto yield_value(Action* action)
        {
            AX_SAFE_RELEASE(_currentAction);
            _currentAction = action;
            AX_SAFE_RETAIN(_currentAction);
            return std::experimental::suspend_always{};
        }

        void return_void() {}
        void unhandled_exception() {}

        void resume()
        {
            handle::from_promise(*this).resume();
        }

    private:
        Action* _currentAction{ nullptr };
    };

    Coroutine() = default;
    Coroutine(const Coroutine&) = delete;
    Coroutine(Coroutine&& rhs) noexcept;
    virtual ~Coroutine();

    Coroutine& operator=(const Coroutine&) = delete;    
    Coroutine& operator=(Coroutine&& rhs) noexcept;

    Action* currentAction() const noexcept;
    bool isDone() const;
    bool moveNext() const;

private:
    Coroutine(handle h) noexcept : _handle{ h } {}

    handle _handle;
};

class AX_DLL ActionCoroutine : public Action
{
public:
    static ActionCoroutine* create(Coroutine&& coroutine);

    virtual bool isDone() const override;
    virtual void step(float dt) override;

public:
    ActionCoroutine() = default;
    virtual ~ActionCoroutine() {}
    bool initWithCoroutine(Coroutine&& coroutine) noexcept;

private:
    AX_DISALLOW_COPY_AND_ASSIGN(ActionCoroutine);

    Coroutine _coroutine;
};

NS_AX_END

#endif      //__ACTION_CCCOROUTINE_ACTION_H__


```
</details>

<details>
    <summary>ActionCoroutine.cpp</summary>

```c++
#include "ActionCoroutine.h"

NS_AX_BEGIN

Coroutine::Coroutine(Coroutine&& rhs) noexcept : _handle(rhs._handle)
{
    rhs._handle = nullptr;
}

Coroutine::~Coroutine()
{
    if (_handle)
    {
        _handle.destroy();
    }
}

Coroutine& Coroutine::operator=(Coroutine&& rhs) noexcept
{
    if (this != std::addressof(rhs))
    {
        _handle = rhs._handle;
        rhs._handle = nullptr;
    }
    return *this;
}

Action* Coroutine::currentAction() const noexcept
{
    return _handle.promise().currentAction();
}

bool Coroutine::isDone() const
{
    return _handle && _handle.done();
}

bool Coroutine::moveNext() const
{
    if (_handle)
    {
        _handle.promise().resume();
        return !_handle.done();
    }
    return false;
}

ActionCoroutine* ActionCoroutine::create(Coroutine&& coroutine)
{
    auto ret = new (std::nothrow) ActionCoroutine();
    if (ret && ret->initWithCoroutine(std::forward<Coroutine>(coroutine)))
    {
        ret->autorelease();
        return ret;
    }
    delete ret;
    return nullptr;
}

bool ActionCoroutine::isDone() const
{
    auto action = _coroutine.currentAction();
    if (action && !action->isDone())
        return false;

    return _coroutine.isDone();
}

void ActionCoroutine::step(float dt)
{
    auto action = _coroutine.currentAction();
    if (action && !action->isDone())
    {
        action->step(dt);
        return;
    }
    _coroutine.moveNext();
}
    
bool ActionCoroutine::initWithCoroutine(Coroutine&& coroutine) noexcept
{
    _coroutine = std::forward<Coroutine>(coroutine);
    return true;
}

NS_AX_END

```
</details>

```c++
Coroutine HelloWorld::testCoroutine()
{
    AXLOG("HelloWorld::testCoroutine() : frame count(%llu)", mFrameCount);
    co_yield DelayTime::create(3.0f);
    AXLOG("HelloWorld::testCoroutine() : frame count(%llu), after 3sec", mFrameCount);
    co_yield nullptr;
    AXLOG("HelloWorld::testCoroutine() : frame count(%llu), next frame", mFrameCount);
}

void HelloWorld::onTouchesBegan(const std::vector<ax::Touch*>& touches, ax::Event* event)
{
    auto action = ActionCoroutine::create(testCoroutine());
    this->runAction(action);

    for (auto&& t : touches)
    {
        //AXLOG("onTouchesBegan detected, X:%f  Y:%f", t->getLocation().x, t->getLocation().y);
    }
}

void HelloWorld::update(float delta)
{
    mFrameCount++;
}
```

image

### PR DESCRIPTION
**Pull request by lich426**
Which branch your pull-request should merge into?

- `dev`: Current 2.x BugFixs or Features
- supported platforms: `ALL`

## Describe your changes
#1471

## Checklist before requesting a review
-  [x] I have performed a self-review of my code.
-  [x] If it is a core feature, I have added thorough tests.
-  [ ] I have checked readme and add important infos to this PR (if it needed).
-  [x] I have added/adapted some tests too.

### CODE REVIEW

File: core/2d/ActionCoroutine.h [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+****************************************************************************/
+
+#ifndef __ACTION_CCCOROUTINE_ACTION_H__
+#define __ACTION_CCCOROUTINE_ACTION_H__
+
+#include "2d/Action.h"
+
+#if (AX_TARGET_PLATFORM == AX_PLATFORM_WIN32 || AX_TARGET_PLATFORM == AX_PLATFORM_WINUWP || \
+     AX_TARGET_PLATFORM == AX_PLATFORM_ANDROID)
+#    if __has_include(<experimental/coroutine>)
```
Comment (halx99) [inline core/2d/ActionCoroutine.h]: I checked the win32 & winuwp have `<coroutine>`, only check android?

File: cmake/Modules/AXConfigDefine.cmake [hunk 2]
```diff
@@ -111,6 +111,10 @@ if (FULL_MSVC)
     add_compile_options(/Bv)
 endif()
 
+if (FULL_MSVC)
```
Comment (halx99) [inline cmake/Modules/AXConfigDefine.cmake]: May don't require additional compiler flags

File: core/2d/ActionCoroutine.h (hunk 1 above)
Comment (lich426) [reply core/2d/ActionCoroutine.h]: I checked the ios & macos.
but linux was not checked

File: core/2d/ActionCoroutine.h [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+#        include <experimental/coroutine>
+#        define COROUTINE_HANDLE(x) std::experimental::coroutine_handle<x>;
+#        define SUSPEND_ALWAYS      std::experimental::suspend_always{};
+#        define AX_SUPPORT_COROUTINE
+#    endif  // __has_include(<experimental/coroutine>)
+
+#elif (AX_TARGET_PLATFORM == AX_PLATFORM_OSX || AX_TARGET_PLATFORM == AX_PLATFORM_IOS || \
```
Comment (halx99) [inline core/2d/ActionCoroutine.h]: follow code may better:
```cpp
#    if __has_include(<coroutine>)
#        include <coroutine>
namespace axstd
{
using suspend_always = std::suspend_always;
template <typename _Ty>
using coroutine_handle = std::coroutine_handle<_Ty>;
}  // namespace axstd
#elif __has_include(<experimental/coroutine>)
// fallback to experimental, currently only for android, in the future may don't required,
// for example: we upgrade ndk in the future releases of axmol
#        include <experimental/coroutine>
namespace axstd
{
using suspend_always = std::experimental::suspend_always;
template <typename _Ty>
using coroutine_handle = std::experimental::coroutine_handle<_Ty>;
}  // namespace axstd
#else
#error This compiler missing c++20 coroutine
#endif
```
then use `axstd::xxx` in `ActionCoroutine` impl code

File: cmake/Modules/AXConfigDefine.cmake (hunk 2 above)
Comment (halx99) [reply cmake/Modules/AXConfigDefine.cmake]: > May don't require additional compiler flags

verified: when include `<coroutine>` in msvc, `/await` don't required

File: tests/cpp-tests/Source/ActionsTest/ActionsTest.cpp [hunk 4]
```diff
@@ -2415,3 +2415,67 @@ std::string Issue18003::subtitle() const
...
+{
+    return "";
+}
+
+Coroutine ActionCoroutineTest::coroutineCallback()
+{
+    auto s = Director::getInstance()->getWinSize();
+
+    auto label1 =
+        Label::createWithTTF(StringUtils::format("First (%llu)", _frameCount), "fonts/Marker Felt.ttf", 16.0f);
```
Comment (halx99) [inline tests/cpp-tests/Source/ActionsTest/ActionsTest.cpp]:
```suggestion
        Label::createWithTTF(StringUtils::format("First (%" PRIu64 ")", _frameCount), "fonts/Marker Felt.ttf", 16.0f);
```

File: tests/cpp-tests/Source/ActionsTest/ActionsTest.cpp [hunk 5]
```diff
@@ -2415,3 +2415,67 @@ std::string Issue18003::subtitle() const
...
+    label1->setPosition(s.width / 4 * 1, s.height / 2);
+    addChild(label1);
+    co_yield DelayTime::create(3.0f);  // delay 3s
+
+    auto label2 =
+        Label::createWithTTF(StringUtils::format("after 3sec (%llu)", _frameCount), "fonts/Marker Felt.ttf", 16.0f);
```
Comment (halx99) [inline tests/cpp-tests/Source/ActionsTest/ActionsTest.cpp]:
```suggestion
        Label::createWithTTF(StringUtils::format("after 3sec (%" PRIu64 ")", _frameCount), "fonts/Marker Felt.ttf", 16.0f);
```

File: tests/cpp-tests/Source/ActionsTest/ActionsTest.cpp [hunk 6]
```diff
@@ -2415,3 +2415,67 @@ std::string Issue18003::subtitle() const
...
+    label2->setPosition(s.width / 4 * 2, s.height / 2);
+    addChild(label2);
+    co_yield nullptr;  // next frame
+
+    auto label3 =
+        Label::createWithTTF(StringUtils::format("next frame (%llu)", _frameCount), "fonts/Marker Felt.ttf", 16.0f);
```
Comment (halx99) [inline tests/cpp-tests/Source/ActionsTest/ActionsTest.cpp]:
```suggestion
        Label::createWithTTF(StringUtils::format("next frame (%" PRIu64 ")", _frameCount), "fonts/Marker Felt.ttf", 16.0f);
```

File: tests/cpp-tests/Source/ActionsTest/ActionsTest.cpp [hunk 7]
```diff
@@ -2415,3 +2415,67 @@ std::string Issue18003::subtitle() const
...
+
+    _frameCount = 1;
+    centerSprites(0);
+
+    auto action = ActionCoroutine::create(coroutineCallback());
+    this->runAction(action);
+
+    auto s = Director::getInstance()->getWinSize();
+    _label =
+        Label::createWithTTF(StringUtils::format("frame count : %llu", _frameCount), "fonts/Marker Felt.ttf", 16.0f);
```
Comment (halx99) [inline tests/cpp-tests/Source/ActionsTest/ActionsTest.cpp]:
```suggestion
        Label::createWithTTF(StringUtils::format("frame count : %" PRIu64 ")", _frameCount), "fonts/Marker Felt.ttf", 16.0f);
```

**Review by halx99 — CHANGES_REQUESTED** [review]

File: tests/cpp-tests/Source/ActionsTest/ActionsTest.cpp [hunk 8]
```diff
@@ -2415,3 +2415,67 @@ std::string Issue18003::subtitle() const
...
+    _label->setPosition(s.width / 2, s.height / 2 + 100);
+    addChild(_label, 1, 1);
+
+    scheduleUpdate();
+}
+
+void ActionCoroutineTest::update(float delta)
+{
+    _frameCount++;
+    _label->setString(StringUtils::format("frame count : %llu", _frameCount));
```
Comment (halx99) [inline tests/cpp-tests/Source/ActionsTest/ActionsTest.cpp]:
```suggestion
    _label->setString(StringUtils::format("frame count : %" PRIu64 ")", _frameCount));
```

File: core/2d/ActionCoroutine.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
+/****************************************************************************
+Copyright (c) 2008-2010 Ricardo Quesada
+Copyright (c) 2010-2012 cocos2d-x.org
+Copyright (c) 2011      Zynga Inc.
+Copyright (c) 2013-2016 Chukong Technologies Inc.
+Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.
+Copyright (c) 2023      lich426 (lich426@gmail.com)
```
Comment (halx99) [inline core/2d/ActionCoroutine.cpp]:
```suggestion
Copyright (c) 2019-present Axmol Engine contributors (see AUTHORS.md).
```

File: core/2d/ActionCoroutine.h [hunk 10]
```diff
@@ -0,0 +1,158 @@
+/****************************************************************************
+Copyright (c) 2008-2010 Ricardo Quesada
+Copyright (c) 2010-2012 cocos2d-x.org
+Copyright (c) 2011      Zynga Inc.
+Copyright (c) 2013-2016 Chukong Technologies Inc.
+Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.
+Copyright (c) 2023      lich426 (lich426@gmail.com)
```
Comment (halx99) [inline core/2d/ActionCoroutine.h]:
```suggestion
Copyright (c) 2019-present Axmol Engine contributors (see AUTHORS.md).
```

**Review by halx99 — CHANGES_REQUESTED** [review]

File: core/2d/ActionCoroutine.h (hunk 10 above)
Comment (halx99) [reply core/2d/ActionCoroutine.h]: I will do:

1. Rename `AUTHORS` -- `AUTHORS-OLD`
2. Add `AUTHORS.md` for new authors list(of axmol) and `lich426 (lich426@gmail.com)` will be in the list

File: templates/cpp-template-default/proj.android/app/build.gradle [hunk 11]
```diff
@@ -25,7 +25,7 @@ android {
             cmake {
                 arguments "-DCMAKE_FIND_ROOT_PATH=", "-DANDROID_STL=c++_shared", "-DANDROID_TOOLCHAIN=clang", "-DANDROID_ARM_NEON=TRUE"
                 arguments.addAll(cmakeOptions)
-                cppFlags "-frtti -fexceptions -fsigned-char"
+                cppFlags "-frtti -fexceptions -fsigned-char -fcoroutines-ts -stdlib=libc++"
```
Comment (halx99) [inline templates/cpp-template-default/proj.android/app/build.gradle]:
```suggestion
                cppFlags "-frtti -fexceptions -fsigned-char"
```

Comment (halx99) [reply templates/cpp-template-default/proj.android/app/build.gradle]: Tested, additional compiler flags also not required for android

**Review by halx99 — APPROVED** [review]

**Pull request merged**